compactor2/src/components/round_info_source/mod.rs

```diff
@@ -73,11 +112,72 @@ impl RoundInfoSource for LevelBasedRoundInfo {
         _partition_info: &PartitionInfo,
         files: &[ParquetFile],
     ) -> Result<Arc<RoundInfo>, DynError> {
+        let start_level = get_start_level(files);
+
+        if self.too_many_files_to_compact(files, start_level) {
+            return Ok(Arc::new(RoundInfo::ManySmallFiles {
+                start_level,
+                max_num_files_to_group: self.max_num_files_per_plan,
+            }));
+        }
+
         let target_level = pick_level(files);
         Ok(Arc::new(RoundInfo::TargetLevel { target_level }))
     }
 }
 
+fn get_start_level(files: &[data_types::ParquetFile]) -> CompactionLevel {
+    // panic if the files are empty
+    assert!(!files.is_empty());
+
+    // Start with initial level
+    // If there are files in  this level, itis the start level
+    // Otherwise repeat until reaching the final level.
+    let mut level = CompactionLevel::Initial;
+    while level != CompactionLevel::Final {
+        if files.iter().any(|f| f.compaction_level == level) {
+            return level;
+        }
+
+        level = level.next();
+    }
+
+    level
+}
+
+fn get_num_overlapped_files(
+    start_level_files: Vec<&ParquetFile>,
+    next_level_files: Vec<&ParquetFile>,
+) -> usize {
+    // min_time and max_time of files in start_level
+    let (min_time, max_time) =
+        start_level_files
+            .iter()
+            .fold((None, None), |(min_time, max_time), f| {
+                let min_time = match min_time {
+                    Some(min_time) => Some(std::cmp::min(min_time, f.min_time)),
+                    None => Some(f.min_time),
+                };
```

Comment:
I think you could also write this more functionally if you wanted:
```suggestion
                let min_time = min_time.map(|v| v.min(f.min_time)).unwrap_or(f.min_time);
```